File: source/directsound/SoundBuffer.cpp
/*
* Copyright (c) 2007 SlimDX Group
* 
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files (the "Software"), to deal
* in the Software without restriction, including without limitation the rights
* to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
* copies of the Software, and to permit persons to whom the Software is
* furnished to do so, subject to the following conditions:
* 
* The above copyright notice and this permission notice shall be included in
* all copies or substantial portions of the Software.
* 
* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
* THE SOFTWARE.
*/
#pragma once

#include <windows.h>
#include <mmreg.h>
#include <dsound.h>

#include "../DataStream.h"
#include "DirectSound.h"
#include "SoundException.h"
#include "WaveFormat.h"
#include "SoundBuffer.h"

using namespace System::Runtime::InteropServices;

namespace SlimDX
{
namespace DirectSound
{
	void BufferDescription::Marshal( DSBUFFERDESC& desc )
	{
		desc.dwSize = sizeof(DSBUFFERDESC);
		desc.dwFlags = static_cast<DWORD>( this->Flags );
		desc.dwBufferBytes = static_cast<DWORD>( this->SizeInBytes );
		desc.dwReserved = 0;

		if( this->Format == nullptr )
			desc.lpwfxFormat = NULL;
		else
			desc.lpwfxFormat = this->Format->InternalPointer;
	}

	SoundBuffer::SoundBuffer( IDirectSoundBuffer* buffer ) : DirectXObject( static_cast<IDirectSoundBuffer8*>( buffer ) )
	{
		if( buffer == NULL )
			throw gcnew ArgumentNullException( "buffer" );

		SetDS8Pointer();
	}

	SoundBuffer::SoundBuffer( DirectSound^ dsound, BufferDescription desc )
	{
		DSBUFFERDESC nativeDesc;
		desc.Marshal( nativeDesc );

		IDirectSoundBuffer* buffer;
		HRESULT hr = dsound->InternalPointer->CreateSoundBuffer( &nativeDesc, &buffer, NULL );
		SoundException::CheckHResult( hr );
		if( FAILED( hr ) )
			throw gcnew SoundException();

		m_Pointer = buffer;
		SetDS8Pointer();
	}

	SoundBuffer::~SoundBuffer()
	{
		if( m_DS8Pointer != NULL )
		{
			m_DS8Pointer->Release();
			m_DS8Pointer = NULL;
		}

		Destruct();
	}

	void SoundBuffer::SetDS8Pointer()
	{
		IDirectSoundBuffer8* buffer;
		HRESULT hr = m_Pointer->QueryInterface( IID_IDirectSoundBuffer8, reinterpret_cast<void**>( &buffer ) );
		
		if( SUCCEEDED( hr ) )
		{
			m_DS8Pointer = buffer;
		}
		else
		{
			m_DS8Pointer = NULL;
		}
	}

	void SoundBuffer::Initialize( DirectSound^ dsound, BufferDescription desc )
	{
		DSBUFFERDESC nativeDesc;
		desc.Marshal( nativeDesc );

		IDirectSoundBuffer* buffer;
		HRESULT hr = dsound->InternalPointer->CreateSoundBuffer( &nativeDesc, &buffer, NULL );
		SoundException::CheckHResult( hr );
		if( FAILED( hr ) )
			throw gcnew SoundException();

		m_Pointer = static_cast<IDirectSoundBuffer8*>( buffer );
		SetDS8Pointer();
	}

	void SoundBuffer::Restore()
	{
		HRESULT hr = m_Pointer->Restore();
		SoundException::CheckHResult( hr );
	}

	void SoundBuffer::Play( int priority, PlayFlags flags )
	{
		HRESULT hr = m_Pointer->Play( 0, priority, static_cast<DWORD>( flags ) );
		SoundException::CheckHResult( hr );
	}

	void SoundBuffer::Stop()
	{
		HRESULT hr = m_Pointer->Stop();
		SoundException::CheckHResult( hr );
	}

	DataStream^ SoundBuffer::Lock( int offset, int sizeBytes, LockFlags flags, [Out] DataStream^% secondPart )
	{
		void* buffer1;
		void* buffer2;
		DWORD size1, size2;

		secondPart = nullptr;

		HRESULT hr = m_Pointer->Lock( offset, sizeBytes, &buffer1, &size1, &buffer2, &size2, static_cast<DWORD>( flags ) );
		SoundException::CheckHResult( hr );
		if( FAILED( hr ) )
			return nullptr;

		DataStream^ stream1 = gcnew DataStream( buffer1, size1, true, true, false );
		if( buffer2 != NULL )
		{
			secondPart = gcnew DataStream( buffer2, size2, true, true, false );
		}

		return stream1;
	}

	void SoundBuffer::Unlock( DataStream^ firstPart, DataStream^ secondPart )
	{
		void* buffer2 = NULL;
		int size2 = 0;

		if( secondPart != nullptr )
		{
			buffer2 = secondPart->RawPointer;
			size2 = static_cast<int>( secondPart->Length );
		}

		HRESULT hr = m_Pointer->Unlock( firstPart->RawPointer, static_cast<int>( firstPart->Length ), buffer2, size2 );
		SoundException::CheckHResult( hr );
	}

	void SoundBuffer::SetFormat( WaveFormatExtended^ format )
	{
		HRESULT hr = m_Pointer->SetFormat( format->InternalPointer );
		SoundException::CheckHResult( hr );
	}

	WaveFormatExtended^ SoundBuffer::GetFormat()
	{
		WAVEFORMATEX* format;
		DWORD size;

		HRESULT hr = m_Pointer->GetFormat( NULL, 0, &size );
		SoundException::CheckHResult( hr );
		if( FAILED( hr ) )
			return nullptr;

		format = reinterpret_cast<WAVEFORMATEX*>( new char[size] );
		hr = m_Pointer->GetFormat( format, size, NULL );
		SoundException::CheckHResult( hr );
		if( FAILED( hr ) )
			return nullptr;

		if( format->wFormatTag == WAVE_FORMAT_EXTENSIBLE )
			return gcnew WaveFormatExtensible( reinterpret_cast<WAVEFORMATEXTENSIBLE*>( format ) );
		else
			return gcnew WaveFormatExtended( format );
	}

	int SoundBuffer::CurrentPlayPosition::get()
	{
		DWORD position = 0;
		HRESULT hr = m_Pointer->GetCurrentPosition( &position, NULL );
		SoundException::CheckHResult( hr );

		return position;
	}
	
	void SoundBuffer::CurrentPlayPosition::set( int value )
	{
		HRESULT hr = m_Pointer->SetCurrentPosition( value );
		SoundException::CheckHResult( hr );
	}

	int SoundBuffer::CurrentWritePosition::get()
	{
		DWORD position = 0;
		HRESULT hr = m_Pointer->GetCurrentPosition( NULL, &position );
		SoundException::CheckHResult( hr );

		return position;
	}
	
	int SoundBuffer::Frequency::get()
	{
		DWORD freq = 0;
		HRESULT hr = m_Pointer->GetFrequency( &freq );
		SoundException::CheckHResult( hr );

		return freq;
	}

	void SoundBuffer::Frequency::set( int value )
	{
		HRESULT hr = m_Pointer->SetFrequency( value );
		SoundException::CheckHResult( hr );
	}

	int SoundBuffer::Volume::get()
	{
		LONG volume = 0;
		HRESULT hr = m_Pointer->GetVolume( &volume );
		SoundException::CheckHResult( hr );

		return volume;
	}

	void SoundBuffer::Volume::set( int value )
	{
		HRESULT hr = m_Pointer->SetVolume( value );
		SoundException::CheckHResult( hr );
	}

	int SoundBuffer::Pan::get()
	{
		LONG pan = 0;
		HRESULT hr = m_Pointer->GetPan( &pan );
		SoundException::CheckHResult( hr );

		return pan;
	}

	void SoundBuffer::Pan::set( int value )
	{
		HRESULT hr = m_Pointer->SetPan( value );
		SoundException::CheckHResult( hr );
	}

	BufferStatus SoundBuffer::Status::get()
	{
		DWORD status = 0;
		HRESULT hr = m_Pointer->GetStatus( &status );
		SoundException::CheckHResult( hr );

		return static_cast<BufferStatus>( status );
	}
}
}

File: SlimDX2/Sources/SlimDX2.DirectSound/Extension/SoundBuffer.cs
// Copyright (c) 2007-2010 SlimDX Group
// 
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
// 
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
// 
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.
using System;

namespace SlimDX2.DirectSound
{
    public partial class SoundBuffer
    {



        /// <unmanaged>HRESULT IDirectSoundBuffer::Play([None] int dwReserved1,[None] int dwPriority,[None] int dwFlags)</unmanaged>
        public SlimDX2.Result Play(int priority, SlimDX2.DirectSound.PlayFlags flags)
        {
            return Play(0, priority, flags);
        }


        /// <summary>	
        /// The Lock method readies all or part of the buffer for a data write and returns pointers to which data can be written.	
        /// </summary>	
        /// <param name="offset"> Offset, in bytes, from the start of the buffer to the point where the lock begins. This parameter is ignored if DSBLOCK_FROMWRITECURSOR is specified in the dwFlags parameter. </param>
        /// <param name="sizeBytes"> Size, in bytes, of the portion of the buffer to lock. The buffer is conceptually circular, so this number can exceed the number of bytes between dwOffset and the end of the buffer. </param>
        /// <param name="flags"> Flags modifying the lock event. The following flags are defined:  ValueDescription DSBLOCK_FROMWRITECURSORStart the lock at the write cursor. The dwOffset parameter is ignored. DSBLOCK_ENTIREBUFFERLock the entire buffer. The dwBytes parameter is ignored.  </param>
        /// <param name="secondPart"> Address of a variable that receives a pointer to the second locked part of the capture buffer. If NULL is returned, the ppvAudioPtr1 parameter points to the entire locked portion of the capture buffer. </param>
        /// <returns>Address of a variable that receives a pointer to the first locked part of the buffer.</returns>
        /// <unmanaged>HRESULT IDirectSoundBuffer::Lock([None] int dwOffset,[None] int dwBytes,[Out] void** ppvAudioPtr1,[Out] int* pdwAudioBytes1,[Out] void** ppvAudioPtr2,[Out, Optional] int* pdwAudioBytes2,[None] int dwFlags)</unmanaged>
        public DataStream Lock(int offset, int sizeBytes, LockFlags flags, out DataStream secondPart)
        {
            IntPtr dataPart1;
            int sizePart1;
            IntPtr dataPart2;
            int sizePart2;
            Lock(offset, sizeBytes, out dataPart1, out sizePart1, out dataPart2, out sizePart2, flags);

            secondPart = null;
            if (dataPart2 != IntPtr.Zero)
                secondPart = new DataStream(dataPart2, sizePart2, true, true);
            
            return new DataStream(dataPart1, sizePart1, true, true);
        }

        /// <summary>	
        /// The Unlock method releases a locked sound buffer.	
        /// </summary>	
        /// <param name="vAudioPtr1Ref"> Address of the value retrieved in the ppvAudioPtr1 parameter of the {{Lock}} method. </param>
        /// <param name="audioBytes1"> Number of bytes written to the portion of the buffer at pvAudioPtr1. See Remarks. </param>
        /// <param name="vAudioPtr2Ref"> Address of the value retrieved in the ppvAudioPtr2 parameter of the IDirectSoundBuffer8::Lock method. </param>
        /// <param name="audioBytes2"> Number of bytes written to the portion of the buffer at pvAudioPtr2. See Remarks. </param>
        /// <returns>No documentation.</returns>
        /// <unmanaged>HRESULT IDirectSoundBuffer::Unlock([In, Buffer] void* pvAudioPtr1,[None] int dwAudioBytes1,[In, Buffer, Optional] void* pvAudioPtr2,[None] int dwAudioBytes2)</unmanaged>
        public SlimDX2.Result Unlock(DataStream dataPart1, DataStream dataPart2)
        {
            if (dataPart2 != null)
                return Unlock(dataPart1.DataPointer, (int)dataPart1.Length, dataPart2.DataPointer, (int)dataPart2.Length);
            else
                return Unlock(dataPart1.DataPointer, (int)dataPart1.Length, IntPtr.Zero, 0);
        }

        /// <summary>
        /// Writes data to the buffer.
        /// </summary>
        /// <returns />
        public void Write<T>(T[] data, int bufferOffset, LockFlags flags) where T : struct
        {
            this.Write<T>(data, 0, 0, bufferOffset, flags);
        }

        /// <summary>
        /// Writes data to the buffer.
        /// </summary>
        /// <returns />
        public void Write<T>(T[] data, int startIndex, int count, int bufferOffset, LockFlags flags) where T : struct
        {
            DataStream stream2 = null;
            //Utilities.CheckArrayBounds(data, startIndex, ref count);
            int bytes = Utilities.SizeOf<T>() * count;
            DataStream stream1 = this.Lock(bufferOffset, bytes, flags, out stream2);
            int count1 = (int)(((long)((int)stream1.Length)) / (Utilities.SizeOf<T>()));
            stream1.WriteRange<T>(data, startIndex, count1);
            if ((stream2 != null) && (count > count1))
            {
                stream2.WriteRange<T>(data, count1 + startIndex, (int)(((long)((int)stream2.Length)) / (Utilities.SizeOf<T>())));
            }
            Unlock(stream1, stream2);
        }
    }
}